src/credentials.cpp

```diff
@@ -90,101 +114,95 @@ namespace preferences
-		assert(lookup.back() == 0);
-		std::string temp = decrypt(credentials, build_key("*.*", get_system_username()));
-		std::string pass = escape(encrypt(key, build_key(server, login)));
-		size_t pos = temp.find(lookup);
-		if(pos == std::string::npos) {
-			while(!temp.empty() && temp.back() == '\xc') {
-				temp.pop_back();
-			}
-			std::copy(lookup.begin(), lookup.end(), std::back_inserter(temp));
-			std::copy(pass.begin(), pass.end(), std::back_inserter(temp));
-			temp.push_back('\xc');
-			credentials = encrypt(temp, build_key("*.*", get_system_username()));
-			std::fill(temp.begin(), temp.end(), '\0');
-			return;
+		auto cred = std::find_if(credentials.begin(), credentials.end(), [&](const login_info& cred) {
+			return cred.server == server && cred.username == login;
+		});
+		if(cred == credentials.end()) {
+			// This is equivalent to emplace_back, but also returns the iterator to the new element
+			cred = credentials.emplace(credentials.end(), login, server, "");
 		}
-		size_t pos2 = temp.find('\xc', pos + lookup.size() - 1);
-		if(pos2 == std::string::npos) {
-			// TODO: Not sure if this is the right thing to do in this case (or if there's even anything that CAN be done in this case)
-			std::copy(pass.begin(), pass.end(), std::back_inserter(temp));
-			temp.push_back('\xc');
-			credentials = encrypt(temp, build_key("*.*", get_system_username()));
-			std::fill(temp.begin(), temp.end(), '\0');
-			return;
-		}
-		temp.replace(pos + lookup.size(), pos2 - pos - lookup.size(), pass);
-		credentials = encrypt(temp, build_key("*.*", get_system_username()));
-		std::fill(temp.begin(), temp.end(), '\0');
+		cred->key = escape(encrypt(key, build_key(server, login)));
 	}
 
 	void load_credentials()
 	{
 		if(!remember_password()) {
 			return;
 		}
-		// Credentials are in a different file, which is a binary blob
-		filesystem::scoped_istream stream = filesystem::istream_file(filesystem::get_credentials_file(), false);
+		clear_credentials();
+		std::string cred_file = filesystem::get_credentials_file();
+		if(!filesystem::file_exists(cred_file)) {
+			return;
+		}
+		filesystem::scoped_istream stream = filesystem::istream_file(cred_file, false);
+		// Credentials file is a binary blob, so use streambuf iterator
 		stream->seekg(0, std::ios::end);
-		credentials.clear();
 		credentials.reserve(stream->tellg());
 		stream->seekg(0, std::ios::beg);
 		stream->clear();
-		credentials.assign(std::istreambuf_iterator<char>(*stream), std::istreambuf_iterator<char>());
+		std::string data((std::istreambuf_iterator<char>(*stream)), (std::istreambuf_iterator<char>()));
+		data = decrypt(data, build_key("*.*", get_system_username()));
+		if(data.empty() || data[0] != CREDENTIAL_SEPARATOR) {
+			ERR_CFG << "Invalid data in credentials file\n";
+			std::fill(data.begin(), data.end(), '\0');
+			return;
+		}
+		for(const std::string elem : utils::split(data, CREDENTIAL_SEPARATOR, utils::REMOVE_EMPTY)) {
+			size_t at = elem.find_first_of('@');
```

Comment:
If = is allowed in usernames, you need to pass a start index for the search:

    size_t eq = elem.find_first_of('=', at + 1);

Otherwise `eq` would point to the = _within_ the username.